### ISSUE TITLE
JSX does not mark `React` import as used

### ISSUE THREAD

**Issue by dallonf** [palantir/tslint#689]
For input: 

```ts
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as config from 'config';

function render(id: string) {
  ReactDOM.render(
    <AppContextWrapper>
      <AssetPage id={id} />
    </AppContextWrapper>,
    document.getElementById('app')
  );
}

render(config.asset);
```

with the `no-unused-variable` rule enabled, throws the warning: `unused variable: 'React'`. Since JSX compiles to `React.createElement`, this is a false positive.

**Comment by dallonf** [comment]
Oops, just noticed https://github.com/palantir/tslint/issues/589 was a closed ticket. I'm not sure I like the suggested solutions though (either manually exclude React from this rule in every single JSX file, or make React a global variable)

**Comment by myitcv** [comment]
@dallonf - agree that turning off the rule for every time you `import` React is cumbersome and defeats the point of a linter.

The second suggestion however is not to make React a global variable; rather it is to use the [global typings](https://github.com/borisyankov/DefinitelyTyped/blob/master/react/react-global.d.ts) definition.

However, you need to be slightly careful. Because, and this is the case for us (I work with @priyatham-aps), we rely on the import of React in order to ensure modules are loaded when required (we use `jspm` and SystemJS). Hence our files now look like this:

```typescript
/// <reference path="../../tsd.d.ts" />

// the above d.ts contains a reference to the global React definition

// this next import ensures that SystemJS loads the module if it hasn't already been loaded
import "react"; 

class MyComponent extends React.Component<{}, {}> {
	render(): JSX.Element {
		return <span>Hello</span>;
	}
}
```

This satisfies the linter, however there is one dangerous side effect: there could be a `.tsx` file in which we forget to `import "react"`. If we do, we won't catch it at compile time, but we might still get bitten at runtime (depending on the load order of files).

So I'm minded to agree that either:

* this rule should be modified to have special knowledge of `React` as an import
* a similar rule is created that has special knowledge of `React` as an import

@adidahiya - thoughts?

**Comment by adidahiya** [comment]
It seems like giving `no-unused-variable` special knowledge of React would get pretty messy -- what if you're using the compiler option `--jsx=preserve`? Then you can't assume that `React` is used unless TSLint also knows about your compiler rules.

Another solution would be to add support for "excludes" to `no-unused-variable` (similar to #522)? But it's unclear if there are truly valid use cases for this besides the `React` variable...

**Comment by myitcv** [comment]
> what if you're using the compiler option `--jsx=preserve`

Hmm, I don't think it actually matters what mode we are in: `React` will always be used, and therefore a reference to it will be required (and the module will need to have been loaded, cf. SystemJS). If you have `--jsx=preserve` then the Babel (or other) transpilation will ultimately rewrite to `React.XYZ` albeit later in the build step.

I think we're actually on safe ground making an exception under the following conditions:

* we are linting a `.tsx` file
* the exception is exactly as follows:

```typescript
import * as React from "react";
``` 

Thoughts?

**Comment by myitcv** [comment]
Sorry, I ignored your suggestion:

> Another solution would be to add support for "excludes" to `no-unused-variable` (similar to #522)? But it's unclear if there are truly valid use cases for this besides the `React` variable...

Agreed, not keen on this - we only have a `React` use case right now, and this use-case is very tightly coupled to decisions that have been made to support `.tsx` in the TypeScript compiler.

**Comment by Pajn** [comment]
If there are jsx syntax in a tsx file React is used so the warning is erroneous. As the walker supports tsx it's already implicitly React specific so I would say the cleanest solution is to consider React used if the walker visits at least one jsx element.

**Comment by adidahiya** [comment]
If we go with that implementation ("if the `no-unused-variable` walker visits at least one JSX element, consider `React` to be a used value in that file") we would have to add the caveat in the docs that we now have a _false negative_ in the case where you erroneously import React but are actually using another JSX-compatible framework (like [Mithril](https://lhorie.github.io/mithril/index.html)).

... perhaps just to be safe we should add a `"react"` option to the `no-unused-variable` rule which enables this functionality.

**Comment by myitcv** [comment]
@adidahiya - great point, I hadn't considered other such frameworks at all.

> ... perhaps just to be safe we should add a `"react"` option to the no-unused-variable rule which enables this functionality.

And just to confirm, by "enables this functionality" are you specifically referring to these two conditions?

* we are linting a `.tsx` file
* the unused variable is defined exactly as follows:

```typescript
import * as React from "react";
```

**Comment by adidahiya** [comment]
Yep, sorry if that wasn't clear -- "enable this functionality" means to _fix_ the false positive we currently report where usage of JSX syntax (only possible in `.tsx` files) signifies an implicit usage of the `React` variable.

**Comment by myitcv** [comment]
Sorry, it's probably my misunderstanding!

I'm just looking to confirm that, under your proposal, with a config that contains:

```json
  "no-unused-variable": [true, "react"],
```

the following will *not* report an error:

```typescript
/// <reference path="tsd.d.ts" />

import * as React from "react";

let x = <span></span>;
console.log(x);
```

where it currently does:

```
app/example.tsx[3, 13]: unused variable: 'React'
```

Reason being the use-case I described [earlier](https://github.com/palantir/tslint/issues/689#issuecomment-143804133)

**Comment by adidahiya** [comment]
@myitcv yep, we are in sync, you understand my proposal correctly

**Comment by adidahiya** [comment]
fixed as of `v3.0.0-dev.2`

**Comment by myitcv** [comment]
FYI - I'm seeing some issues with this rule in our code base... for some reason we're not getting 'visit' callbacks for every JSX element. I'm investigating.

**Comment by adidahiya** [comment]
@myitcv I think you need to also visit `JsxSelfClosingElement`s. that's why I do in the whitespace rule to skip JSX. should be a simple fix

**Comment by adidahiya** [comment]
@myitcv fixed here: https://github.com/palantir/tslint/pull/776

**Comment by donaldpipowitch** [comment]
Should this work?

````
{
  "rules": {
    "no-unused-variable":  [
      true,
      "react"
    ]
  }
}
````

It still throws here:

````
import React from 'react';
import { render } from 'react-dom';
import { Router, Route, IndexRedirect, hashHistory } from 'react-router';
import Page from './page/component';
import GistsContainer from './gists/container';
import RepositoriesContainer from './repositories/container';

const DEFAULT_USER = 'octocat';

render(
  <Router history={hashHistory}>
    <Route path="/">
      <IndexRedirect to={`/users/${DEFAULT_USER}/gists`} />
      <Route path="users/:user" component={Page}>
        <Route path="gists" component={GistsContainer} />
        <Route path="repositories" component={RepositoriesContainer} />
      </Route>
    </Route>
  </Router>,
  document.getElementById('app')
);
````

I use `"tslint": "^3.7.1",`.

**Comment by JKillian** [comment]
@donaldpipowitch It should, but it's currently a bug: #893

**Comment by donaldpipowitch** [comment]
Okay, thank you. I'll follow the other bug.
